File: server/src/main/java/com/fluffy/exam/domain/QuestionOption.java
package com.fluffy.exam.domain;

import com.fluffy.global.exception.BadRequestException;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
public class QuestionOption {

    private static final int MAX_TEXT_LENGTH = 200;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String text;

    @Column(nullable = false)
    private boolean isCorrect;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "question_id")
    private Question question;

    public QuestionOption(String text, boolean isCorrect) {
        this(null, text, isCorrect);
    }

    public QuestionOption(Long id, String text, boolean isCorrect) {
        validate(text);

        this.id = id;
        this.text = text;
        this.isCorrect = isCorrect;
    }

    public void validate(String text) {
        validateTextLength(text);
    }

    public void validateTextLength(String text) {
        if (text.isBlank() || text.length() > MAX_TEXT_LENGTH) {
            throw new BadRequestException("질문 옵션의 길이는 1~%d자 이어야 합니다.".formatted(MAX_TEXT_LENGTH));
        }
    }

    public void updateQuestion(Question question) {
        this.question = question;
    }
}

File: server/src/main/kotlin/com/fluffy/exam/domain/QuestionOption.kt
package com.fluffy.exam.domain

import com.fluffy.global.exception.BadRequestException
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne

@Entity
class QuestionOption(
    text: String,
    isCorrect: Boolean,
    id: Long = 0,
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = id
        protected set

    @field:Column(nullable = false)
    var text: String = text
        protected set

    @field:Column(nullable = false)
    var isCorrect: Boolean = isCorrect
        protected set

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "question_id", nullable = false)
    var question: Question? = null

    companion object {

        private const val MAX_TEXT_LENGTH = 200
    }

    init {
        validateTextNotBlank()
        validateTextLength()
    }

    fun updateQuestion(question: Question) {
        this.question = question
    }

    private fun validateTextNotBlank() {
        if (text.isBlank()) {
            throw BadRequestException("문제 선택지의 내용은 비어있을 수 없습니다.")
        }
    }

    private fun validateTextLength() {
        if (text.length > MAX_TEXT_LENGTH) {
            throw BadRequestException("문제 선택지의 내용은 ${MAX_TEXT_LENGTH}자 이하여야 합니다.")
        }
    }
}